packages/telemetry/src/context-aware-slog.js

```diff
@@ -0,0 +1,475 @@
+    );
+
+  const loggerProvider = new LoggerProvider({
+    resource: new Resource(getResourceAttributes(options)),
+  });
+  loggerProvider.addLogRecordProcessor(
+    new SimpleLogRecordProcessor(new OTLPLogExporter({ keepAlive: true })),
+  );
+
+  logs.setGlobalLoggerProvider(loggerProvider);
+  const logger = logs.getLogger('default');
+
+  const db = await getDatabaseInstance();
+
+  /** @type Array<Context | null> */
+  let [
+    blockContext,
+    crankContext,
+    initContext,
+    lastPersistedTriggerContext,
+    replayContext,
+    triggerContext,
+  ] = [null, null, null, null, null, null];
+
+  /**
+   * @param {Context} context
+   */
+  const persistContext = context => {
+    lastPersistedTriggerContext = context;
+    db.persistContext(context);
+  };
+
+  const restoreContext = () => {
+    if (lastPersistedTriggerContext) return lastPersistedTriggerContext;
+    return db.restoreContext();
+  };
+
+  /**
+   * @param {Slog} slog
+   */
+  const slogSender = ({ monotime, time: timestamp, ...body }) => {
+    const finalBody = { ...body };
+
+    /** @type {LogAttributes} */
+    let logAttributes = {
+      'chain-id': String(CHAIN_ID),
+      'process.uptime': monotime,
+      timestamp,
+    };
+
+    /**
+     * Add any before report operations here
+     * like setting context data
+     */
+    switch (body.type) {
+      case SLOG_TYPES.CONSOLE: {
+        delete finalBody.crankNum;
+        delete finalBody.deliveryNum;
+
+        break;
+      }
+      case SLOG_TYPES.CLIST: {
+        assert(!!crankContext);
+        crankContext['crank.vatID'] = finalBody.vatID;
+        break;
+      }
+      case SLOG_TYPES.CRANK.START: {
+        crankContext = {
+          'crank.num': finalBody.crankNum,
+          'crank.type': finalBody.crankType,
+        };
+        break;
+      }
+      case SLOG_TYPES.DELIVER: {
+        if (replayContext) {
+          assert(finalBody.replay);
+          replayContext = {
+            ...replayContext,
+            'crank.deliveryNum': finalBody.deliveryNum,
+            'crank.vatID': finalBody.vatID,
+          };
+        } else {
+          assert(!!crankContext);
+          crankContext = {
+            ...crankContext,
+            'crank.deliveryNum': finalBody.deliveryNum,
+            'crank.vatID': finalBody.vatID,
+          };
+        }
+
+        delete finalBody.deliveryNum;
+        delete finalBody.replay;
+
+        break;
+      }
+      case SLOG_TYPES.DELIVER_RESULT: {
+        delete finalBody.deliveryNum;
+        delete finalBody.replay;
+
+        break;
+      }
+      case SLOG_TYPES.KERNEL.INIT.START: {
+        initContext = { init: true };
+        break;
+      }
+      case SLOG_TYPES.REPLAY.START: {
+        replayContext = { replay: true };
+        break;
+      }
+      // eslint-disable-next-line no-restricted-syntax
+      case SLOG_TYPES.RUN.START: {
+        if (!(triggerContext || finalBody.runNum === 0)) {
```

Comment:
I find the negation confusing.

```suggestion
        if (!triggerContext && finalBody.runNum !== 0) {
```